crc/api/user.py

```diff
@@ -10,29 +10,113 @@
 """
-def verify_token(token):
-    failure_error = ApiError("invalid_token", "Unable to decode the token you provided.  Please re-authenticate", status_code=403)
-    if (not 'PRODUCTION' in app.config or not app.config['PRODUCTION']) and token == app.config["SWAGGER_AUTH_KEY"]:
+
+
+def verify_token(token=None):
+    """
+        Verifies the token for the user (if provided). If in production environment and token is not provided,
+        gets user from the SSO headers and returns their token.
+
+        Args:
+            token: Optional[str]
+
+        Returns:
+            token: str
+
+        Raises:
+            ApiError.  If not on production and token is not valid, returns an 'invalid_token' 403 error.
+            If on production and user is not authenticated, returns a 'no_user' 403 error.
+   """
+
+    failure_error = ApiError("invalid_token", "Unable to decode the token you provided.  Please re-authenticate",
+                             status_code=403)
+
+    if not _is_production():
         g.user = UserModel.query.first()
         token = g.user.encode_auth_token()
 
-    try:
-        token_info = UserModel.decode_auth_token(token)
-        g.user = UserModel.query.filter_by(uid=token_info['sub']).first()
-    except:
-        raise failure_error
-    if g.user is not None:
-        return token_info
+    if token:
+        try:
+            token_info = UserModel.decode_auth_token(token)
+            g.user = UserModel.query.filter_by(uid=token_info['sub']).first()
+        except:
+            raise failure_error
+        if g.user is not None:
+            return token_info
+        else:
+            raise failure_error
+
+    # If there's no token and we're in production, get the user from the SSO headers and return their token
+    if not token and _is_production():
+        uid = _get_request_uid(request)
+
+        if uid is not None:
+            db_user = UserModel.query.filter_by(uid=uid).first()
+
+            if db_user is not None:
+                g.user = db_user
+                token = g.user.encode_auth_token().decode()
+                token_info = UserModel.decode_auth_token(token)
+                return token_info
+
+            else:
+                raise ApiError("no_user", "User not found. Please login via the frontend app before accessing this feature.",
+                         status_code=403)
+
+
+def verify_token_admin(token=None):
+    """
+        Verifies the token for the user (if provided) in non-production environment. If in production environment,
+        checks that the user is in the list of authorized admins
+
+        Args:
+            token: Optional[str]
+
+        Returns:
+            token: str
+   """
+
+    # If this is production, check that the user is in the list of admins
+    if _is_production():
+        uid = _get_request_uid(request)
+
+        if uid is not None and uid in app.config['ADMIN_UIDS']:
```

Comment:
We will want to refactor this at somepoint to use permissions and roles, and have an interface where admins can make other admins or grant permissions, but this is an elegant in-between point.